README.md
# mod-service-interaction
Copyright (C) 2018-2019 The Open Library Foundation

This software is distributed under the terms of the Apache License,
Version 2.0. See the file "[LICENSE](LICENSE)" for more information.

# Introduction - developers looking to enhance the resources that mod-service-interaction provides

Mod-Service-Interaction is a FOLIO Backend Module for cross-app connectivity.

Mod-Service-Interaction is currently just storage for the dashboard, but since it handles things such as making multi-interface okapi calls and consolidating the returned values, this module represents an opportunity for other such operations, particularly in an ERM context, to have a home in future.

Developers looking to access the services exposed by mod-service-interaction can find more information in the following documentation:
[See the documentation](https://wiki.folio.org/display/ERM/Dashboard+Documentation)

## ModuleDescriptor
https://github.com/folio-org/mod-service-interaction/blob/master/service/src/main/okapi/ModuleDescriptor-template.json

# For module developers looking to extend or modify the resources presented by this module

This is the main starter repository for the Grails-based OLF - ERM backend modules.

- [Getting started](service/docs/getting-started.md "Getting started")

# For ops folks trying to deploy the module

A sample k8s deployment and service resource description can be [found in the scripts directory](https://github.com/folio-org/mod-service-interaction/blob/master/scripts/k8s_deployment_template.yaml)

Most importantly, the module requires a number of ENV settings which are different to the RMB defaults
- OKAPI_SERVICE_PORT - port number for okapi
- OKAPI_SERVICE_HOST - host for okapi - in K8S [namespace.]hostname or just hostname if you are running the pod in the same namespace as okapi

## Additional information

### Issue tracker

See project [ERM](https://issues.folio.org/projects/ERM)
at the [FOLIO issue tracker](https://dev.folio.org/guidelines/issue-tracker/).

### Other documentation

Other [modules](https://dev.folio.org/source-code/#server-side) are described,
with further FOLIO Developer documentation at [dev.folio.org](https://dev.folio.org/)


## Running using grails run-app with the vagrant-db profile

    grails -Dgrails.env=vagrant-db run-app


## Initial Setup

Most developers will run some variant of the following commands the first time through

### In window #1

Start the vagrant image up from the project root

    vagrant destroy
    vagrant up

Sometimes okapi does not start cleanly in the vagrant image - you can check this with

    vagrant ssh

then once logged in

    docker ps

should list running images - if no processes are listed, you will need to restart okapi (In the vagrant image) with

    sudo su - root
    service okapi stop
    service okapi start

Finish the part off with

    tail -f /var/log/folio/okapi/okapi.log

### In window #2

Build and run mod-mod-service-int stand alone

    cd service
    grails war
    ../scripts/run_external_reg.sh

### In window #3

Register the module

  cd scripts
  ./register_and_enable.sh


### In window #4

Run up a stripes platform containing [ui-dashboard](https://github.com/folio-org/ui-dashboard)

---
**NOTE**

platform-erm does not yet contain ui-dashboard, it will soon.

---

This section is run in a local setup, not from any particular checked out project, YMMV

    cd ../platform/stripes/platform-erm
    stripes serve ./stripes.config.js --has-all-perms



You should get back

Waiting for webpack to build...
Listening at http://localhost:3000

and then be able to access the app

  



NEWS.md
## 2.2.1 2023-02-20
  * ERM-2433 Bumped dependencies of postgres, opencsv, web-toolkit and grails-okapi
    * Addec migrations to handle updates to grails-okapi and web-toolkit
  * Added migrations for better number generator generator names/codes (Code changes only for those not in use in production already)
  * Number generator names bootstrapped into the system were too generic, changes to bootstrapped data and migrations for existing data made
  * Fix for initialiseDefaultSequence, wasn't saving thanks to non-nullable name field

## 2.1.0 2023-01-10
  * Fix for CheckDigitAlgo

## 2.0.0 2022-10-25
  * ERM-2312 Managed Dashboards: backend model
    * Changed domain model to allow for multiple dashboards per user and multiple users per dashboard
    * Changed endpoints to reflect this (Breaking change)
  * Number generator
    * Refactors
    * Added configuration for user sequences
    * NextValue defaults to 1
    * Added `enabled` and `description` fields
    * Added `name` field

## 1.1.0 2022-06-29
  * ERM-2134 Service Interaction - mod-service-interaction lacks memory limit in launch descriptor
  * ERM-2071 mod-service-interaction Grails wrapper SAXParseException
  * Number generator
    * Added Number Generator domain classes
    * Number Generator endpoints and services

## 1.0.0 2021-06-15
* ERM-1740: Method to refresh WidgetTypes from scratch
* ERM-1738: Support health check endpoint for mod-service-interaction
* ERM-1696: Support match type search in SimpleSearch Widgets
* ERM-1685: Support "Link" result values
* ERM-1651/ERM-1652/ERM-1653: Support "Array" display values
* ERM-1650: Add unique indexes for refdata tables
* ERM-1643: Manage tenant widget definitions for different applications
* ERM-1580: Added weight to WidgetInstances on dashboard
* ERM-1579: WidgetDef tweaks, added UUID valueType
* ERM-1562: Formatting of bootstrapped data, added licenses definition and "Enum" valueType
* ERM-1546: Added some more structure to widgetType
* ERM-1529: Initial setup, added domain classes, controllers and initial widget data

scripts/secure_enable.sh
## This script logs into OKAPI and then registers and enables the module
BASEDIR=$(dirname "$0")

MODULE_PORT=8078

# This script generates the module descriptor for mod-rs and posts it to a secured OKAPI control interface
# the script is controlled by a ~/.okapirc file where you need to specify the supertenant username (ST_UN)
# supertnent password (ST_PW) and the OKAPI_URL (For the rancher-desktop install, likely http://localhost:30100)

if [ -f .okapirc ]; then
  . .okapirc
elif [ -f $HOME/.okapirc ]; then
  . $HOME/.okapirc
else
  echo You must configure \$HOME/.okapirc
  echo export IS_SECURE_SUPERTENANT=Y
  echo export ST_UN=sysadm
  echo export ST_PW=PASSWORD_FROM_LOCAL_okapi_commander_cfg.json
  echo export OKAPI_URL=http://localhost:30100
  exit 0
fi

echo $BASEDIR
pushd "$BASEDIR/../service"

DIR="$BASEDIR/../"

echo "Using directory: $DIR"

# Check for decriptor target directory.

DESCRIPTORDIR="../service/build/resources/main/okapi"

if [ ! -d "$DESCRIPTORDIR" ]; then
    echo "No descriptors found. Let's try building them."
    
    ./gradlew generateDescriptors
fi

# DEP_DESC=`cat ${DESCRIPTORDIR}/DeploymentDescriptor.json | jq -c ".url=\"$2\""`
DEP_DESC=`cat ${DESCRIPTORDIR}/DeploymentDescriptor.json | jq -c ".url=\"http://192.168.5.2:$MODULE_PORT/\""`
SVC_ID=`echo $DEP_DESC | jq -rc '.srvcId'`
INS_ID=`echo $DEP_DESC | jq -rc '.instId'`

AUTH_TOKEN=`../scripts/okapi-login -u $ST_UN -p $ST_PW -t supertenant`

echo "AUTH_TOKEN: $AUTH_TOKEN"
echo "SUPERTENANT UN: $ST_UN, PASSWORD: $ST_PW"

ENABLE_DOC=`echo $DEP_DESC | jq -c '[{id: .srvcId, action: "enable"}]'`
echo "Enable service - enable doc is $ENABLE_DOC"

curl -XPOST -H "X-Okapi-Token: $AUTH_TOKEN" "${OKAPI_URL}/_/proxy/tenants/$TENANT/install?tenantParameters=loadSample%3Dtest,loadReference%3Dother" -d "$ENABLE_DOC"
echo

echo Done

scripts/run_rancher_desktop.sh
#!/bin/bash
MODULE_PORT=8078
MODULE_NAME=mod-service-interaction

source ~/.okapirc

# This script is for executing a double check that the endpoint calls that work directly
# also work via okapi.

# see https://docs.spring.io/spring-boot/docs/current/reference/html/boot-features-external-config.html for info on overriding 
# spring boot app config on the command line

# This script should be run from the mod-rs/service directory after a "grails prod war" build completes

the_jar_file=`ls build/libs/$MODULE_NAME*.jar | tail -n 1`

echo Start mod-rs in external-register mode - Jar will be $the_jar_file

# curl --header "X-Okapi-Tenant: diku" http://localhost:9130/content -X GET

# THis DOES work as expected however - 
java -Djava.net.preferIPv4Stack=true -Dserver.port=$MODULE_PORT -jar $the_jar_file -Xmx1G --grails.server.host=10.0.2.2 --dataSource.username=okapi_1 --dataSource.password=okapi_1 --dataSource.url=jdbc:postgresql://localhost:30101/okapi_modules


scripts/secure_register.sh
## This script logs into OKAPI and then registers and enables the module
BASEDIR=$(dirname "$0")

MODULE_PORT=8078

# This script generates the module descriptor for mod-rs and posts it to a secured OKAPI control interface
# the script is controlled by a ~/.okapirc file where you need to specify the supertenant username (ST_UN)
# supertnent password (ST_PW) and the OKAPI_URL (For the rancher-desktop install, likely http://localhost:30100)

if [ -f .okapirc ]; then
  . .okapirc
elif [ -f $HOME/.okapirc ]; then
  . $HOME/.okapirc
else
  echo You must configure \$HOME/.okapirc
  echo export IS_SECURE_SUPERTENANT=Y
  echo export ST_UN=sysadm
  echo export ST_PW=PASSWORD_FROM_LOCAL_okapi_commander_cfg.json
  echo export OKAPI_URL=http://localhost:30100
  exit 0
fi

echo $BASEDIR
pushd "$BASEDIR/../service"

DIR="$BASEDIR/../"

echo "Using directory: $DIR"

# Check for decriptor target directory.

DESCRIPTORDIR="../service/build/resources/main/okapi"

if [ ! -d "$DESCRIPTORDIR" ]; then
    echo "No descriptors found. Let's try building them."
    
    ./gradlew generateDescriptors
fi

# DEP_DESC=`cat ${DESCRIPTORDIR}/DeploymentDescriptor.json | jq -c ".url=\"$2\""`
DEP_DESC=`cat ${DESCRIPTORDIR}/DeploymentDescriptor.json | jq -c ".url=\"http://192.168.5.2:$MODULE_PORT/\""`
SVC_ID=`echo $DEP_DESC | jq -rc '.srvcId'`
INS_ID=`echo $DEP_DESC | jq -rc '.instId'`

AUTH_TOKEN=`../scripts/okapi-login -u $ST_UN -p $ST_PW -t supertenant`

echo "AUTH_TOKEN: $AUTH_TOKEN"
echo "SUPERTENANT UN: $ST_UN, PASSWORD: $ST_PW"

echo Remove any existing module ${SVC_ID}/${INS_ID}
echo Waiting for curl -XDELETE "${OKAPI_URL}/_/proxy/tenants/${TENANT_NAME}/modules/${SVC_ID}"
curl -XDELETE -H "X-Okapi-Token: $AUTH_TOKEN" "${OKAPI_URL}/_/proxy/tenants/${TENANT_NAME}/modules/${SVC_ID}"
echo

echo Waiting for curl -XDELETE -H "X-Okapi-Token: $AUTH_TOKEN" "${OKAPI_URL}/_/discovery/modules/${SVC_ID}/${INS_ID}"
curl -XDELETE -H "X-Okapi-Token: $AUTH_TOKEN" "${OKAPI_URL}/_/discovery/modules/${SVC_ID}/${INS_ID}"
echo

echo Waiting for curl -XDELETE -H "X-Okapi-Token: $AUTH_TOKEN" "${OKAPI_URL}/_/proxy/modules/${SVC_ID}"
curl -XDELETE -H "X-Okapi-Token: $AUTH_TOKEN" "${OKAPI_URL}/_/proxy/modules/${SVC_ID}"
echo

# ./gradlew clean generateDescriptors
echo Install latest module ${SVC_ID}/${INS_ID} 
curl -XPOST -H "X-Okapi-Token: $AUTH_TOKEN" ${OKAPI_URL}/_/proxy/modules -d @"${DESCRIPTORDIR}/ModuleDescriptor.json"
echo

echo -e "\n\nPOSTING DEPLOYMENT DESCRIPTOR:"
curl -XPOST -H "X-Okapi-Token: $AUTH_TOKEN" "${OKAPI_URL}/_/discovery/modules" -d "$DEP_DESC"
echo

popd